### ISSUE TITLE
speedup 'smartypants'

### ISSUE THREAD

**Issue by tripodsan** [adobe/helix-pipeline#335]
the typographics quotes formatter of #27 adds quite some performance penalty even for text that does not contain quotes.

**Comment by ramboz** [comment]
I actually wonder on the value of such a formatter in a system that is supposed to support proper internationalization. Smartypants only makes sense on a latin alphabet and introduces a bias towards English in the first place. Quotes for instance are different in English, German and French… and the proposed curly ones only work properly in English.

I'd even argue that proper quotes in HTML should be `<q>` and left to the browser to render based on the locale (or CSS styling). Various dashes also have either different meanings or renderings in various scripts. I'd rather let this be a project-level enhancement via some hooks rather than in the default pipeline which introduces a performance overhead and unnecessary bias towards English on all pages.

**Comment by trieloff** [comment]
Smartypants does following things:

- fix "quotes"
- fix ... ellipsises
- fix -- dashes

Having good typographic practices by default is important to create great looking sites out of the box and requiring a project to enable it will lead to inferior sites and inflated project budgets.

I do agree that `<q>` would be an improvement (and not too hard to build) and that we should have language-specific quotes, too.

**Comment by ramboz** [comment]
> Having good typographic practices by default is important to create great looking sites out of the box 

@trieloff I agree, but let's make sure we don't necessarily introduce an opinionated bias towards English or latin scripts. If we make sure to at least use proper `<q>`/`<blockquote>` for quotes, then we can do CSS styling on top for each locale.

It is a bit more tricky with the ellipsis and the various dashes though, as you can't style them directly.

**Comment by trieloff** [comment]
`blockquote` is already handled correctly, and I'm fine with adding `<q>` (although this is harder than just fixing the quote style, as it requires a new node type)

Having language-dependent quotes would require a language detection step (although that should be relatively easy to build), but we cannot rely on CSS to be there. So replacing the quotes is ok, removing them isn't.

**Comment by ramboz** [comment]
> we cannot rely on CSS to be there

The HTML5 spec usually requires you to define the `lang` for page, and you can easily do the quote styling based on that.

Something like:
```css
html[lang="en"] q { quotes: '“' '”' "‘" "’"; }
html[lang="fr"] q { quotes: "«" "»" "‹" "›"; }
html[lang="de"] q { quotes: "„" "“" "‚" "‘"; }
…
q::before { content: open-quote; }
q:after { content: close-quote; }
```

You'd also probably style the `blockquote` and `cite` elements as well.

> removing them isn't.

👍

**Comment by trieloff** [comment]
What I meant with "we cannot rely on CSS to be there" is "I do not want to force users of Helix to have mandatory CSS, JavaScript or any other client-side code.

**Comment by trieloff** [comment]
I think each of these should be a separate issue (we got sidetracked quite a bit from the original intent of just speeding this stuff up).

We can expose the hook when people ask for the hook ;)
